File: backend/app/models/geo.py
"""GEO/AEO models — AI visibility tracking and AI-readiness auditing."""
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class GeoKeyword(Base):
    """A keyword selected by the user for GEO/AEO tracking."""

    __tablename__ = "geo_keywords"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    keyword: Mapped[str] = mapped_column(String(500), nullable=False)
    # semantic | topvisor | manual
    source: Mapped[str] = mapped_column(String(20), nullable=False, default="manual")
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class GeoScanResult(Base):
    """Result of one AI-model check for one keyword."""

    __tablename__ = "geo_scan_results"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    keyword_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("geo_keywords.id", ondelete="CASCADE"), nullable=False, index=True
    )
    # e.g. "perplexity/llama-3.1-sonar-small-128k-online"
    ai_model: Mapped[str] = mapped_column(String(120), nullable=False)
    mentioned: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    # first | middle | end
    mention_position: Mapped[str | None] = mapped_column(String(20), nullable=True)
    # positive | neutral | negative
    sentiment: Mapped[str | None] = mapped_column(String(20), nullable=True)
    # URLs cited by the model (Perplexity citations field)
    sources_json: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    # other domains mentioned in the response
    competitor_domains_json: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    # first 300 chars of AI response
    response_snippet: Mapped[str | None] = mapped_column(Text, nullable=True)
    scanned_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, index=True)


class AiReadinessAudit(Base):
    """AI-readiness audit for a project (robots, llms.txt, E-E-A-T, freshness)."""

    __tablename__ = "ai_readiness_audits"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    # AI bot blocking
    blocked_bots_json: Mapped[list] = mapped_column(JSONB, nullable=False, default=list)
    cloudflare_detected: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    # llms.txt
    has_llms_txt: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    llms_txt_content: Mapped[str | None] = mapped_column(Text, nullable=True)
    # E-E-A-T basics
    has_about_page: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    has_author_page: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    # Freshness: {url: {last_updated, age_days, status}}
    pages_freshness_json: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    # Overall score 0-100
    ai_readiness_score: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    # Full raw audit data for detailed display
    audit_json: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, index=True)

File: frontend/src/api/geo.ts
import { api } from './client'

export interface GeoKeyword {
  id: string
  keyword: string
  source: 'semantic' | 'topvisor' | 'manual'
}

export interface GeoScanCellResult {
  mentioned: boolean
  position: 'first' | 'middle' | 'end' | null
  sentiment: 'positive' | 'neutral' | 'negative' | null
  snippet: string | null
  competitor_domains: string[]
  scanned_at: string
}

export interface GeoMatrixRow {
  keyword_id: string
  keyword: string
  source: string
  results: Record<string, GeoScanCellResult>  // model → result
}

export interface GeoResultsResponse {
  ai_visibility_score: number | null
  rows: GeoMatrixRow[]
  top_competitors: { domain: string; count: number }[]
}

export interface AiReadinessAudit {
  id: string
  ai_readiness_score: number
  blocked_bots: string[]
  cloudflare_detected: boolean
  has_llms_txt: boolean
  llms_txt_content: string | null
  has_about_page: boolean
  has_author_page: boolean
  pages_freshness: Record<string, { last_updated: string | null; age_days: number | null; status: string }>
  audit_json: Record<string, unknown>
  created_at: string
}

export interface GeoModel {
  id: string
  name: string
  is_default: boolean
}

export const geoApi = {
  // Keywords
  listKeywords: (projectId: string): Promise<GeoKeyword[]> =>
    api.get(`/projects/${projectId}/geo/keywords`).then((r) => r.data),

  addKeywords: (projectId: string, keywords: { keyword: string; source?: string }[]): Promise<GeoKeyword[]> =>
    api.post(`/projects/${projectId}/geo/keywords`, keywords).then((r) => r.data),

  deleteKeyword: (projectId: string, kwId: string): Promise<void> =>
    api.delete(`/projects/${projectId}/geo/keywords/${kwId}`).then(() => undefined),

  // Models
  listModels: (): Promise<GeoModel[]> =>
    api.get('/geo/models').then((r) => r.data),

  // Scan
  startScan: (
    projectId: string,
    kwIds: string[],
    models: string[]
  ): Promise<{ task_id: string; status: string }> =>
    api.post(`/projects/${projectId}/geo/scan`, { keyword_ids: kwIds, models }).then((r) => r.data),

  getResults: (projectId: string): Promise<GeoResultsResponse> =>
    api.get(`/projects/${projectId}/geo/results`).then((r) => r.data),

  // Audit
  runAudit: (projectId: string): Promise<{ task_id: string; status: string }> =>
    api.post(`/projects/${projectId}/geo/audit/run`).then((r) => r.data),

  getAudit: (projectId: string): Promise<AiReadinessAudit | null> =>
    api.get(`/projects/${projectId}/geo/audit`).then((r) => r.data),

  getLlmsTxt: (projectId: string): Promise<{ content: string; filename: string }> =>
    api.get(`/projects/${projectId}/geo/audit/llms-txt`).then((r) => r.data),
}
